torch/_higher_order_ops/flex_attention.py

```diff
@@ -125,31 +125,45 @@ def math_attention(
     """
     working_precision = torch.float64 if query.dtype == torch.float64 else torch.float32
 
+    query = query.view(
+        query.size(0), key.size(1), -1, query.size(-2), query.size(-1)
+    )  # [B, Hq, Mq, D] -> [B, Hkv, G, Mq, D], G=Hq//Hkv
+    key = torch.unsqueeze(key, 2)  # [B, Hkv, Mkv, D] -> [B, Hkv, 1, Mkv, D]
+    value = torch.unsqueeze(value, 2)  # [B, Hkv, Mkv, D] -> [B, Hkv, 1, Mkv, D]
+
     scores = (query @ key.transpose(-2, -1)).to(dtype=working_precision)
 
     b = torch.arange(0, scores.size(0), device=scores.device)
-    h = torch.arange(0, scores.size(1), device=scores.device)
-    m = torch.arange(0, scores.size(2), device=scores.device)
-    n = torch.arange(0, scores.size(3), device=scores.device)
+    hkv = torch.arange(0, scores.size(1), device=scores.device)
+    g = torch.arange(0, scores.size(2), device=scores.device)
+    hq = hkv[:, None] * scores.size(2) + g[None, :]
+    m = torch.arange(0, scores.size(-2), device=scores.device)
+    n = torch.arange(0, scores.size(-1), device=scores.device)
 
     in_dim_buffers = (None,) * len(other_buffers)
     score_mod = torch.vmap(score_mod, in_dims=(0, None, None, None, 0) + in_dim_buffers)
     score_mod = torch.vmap(score_mod, in_dims=(0, None, None, 0, None) + in_dim_buffers)
     score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
+    score_mod = torch.vmap(score_mod, in_dims=(0, None, 0, None, None) + in_dim_buffers)
```

Comment:
I think you might have mentioned, but why does this need to be done? Why can't the head index just represent the query head?

Reply:
The query head dimension is reshaped into (Hkv, G) to allow K, V broadcast along G dimension. From `score_mod` 's perspective, head index does represent query head. That why both Hkv & G dimension are mapped to hq arg of `score_mod`.